### PR TITLE
Adds InitiateAuth, ConfirmSignup operation

### PR DESCRIPTION
**Pull request by idkjs**
There is some stuff in here to look at. Pulling just to show you and get some feedback. The most of the methods almost work. ConfirmSignUp works, rest need to get access to a token which i have not figured out how to do.

Also, added the operation types to a type in Types.re which gives a bs variant to string to restrict putting in kinds that dont exists, misspelled.

Also did that for `AuthFlow` types but have not implemented until i figure out what we are using for authflow types.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -25,3 +25,8 @@ lib/bs
 .merlin
 .bsb.lock
 node_modules
+src/InitiatingAuth.re
+src/**/*.bs.js
```
Comment (pckilgore) [inline .gitignore]:
```suggestion
```
I agree with this: https://reasonml.github.io/docs/en/converting-from-js#tips

File: .gitignore [hunk 2]
```diff
@@ -25,3 +25,8 @@ lib/bs
 .merlin
 .bsb.lock
 node_modules
+src/InitiatingAuth.re
```
Comment (pckilgore) [inline .gitignore]: I suggest making more targeted use of `git add` and `git rm` for local work rather than adding this sort of thing to a global ignore file.  You do lose some shortcuts (`git add .`, `git commit -A ...`) but that pain is nothing compared to losing a bunch of work because of an overly-ambitious ignore file, trust me :cry:.

File: bsconfig.json [hunk 3]
```diff
@@ -1,18 +1,27 @@
...
-    "module": "commonjs",
-    "in-source": true
-  },
-  "suffix": ".bs.js",
-  "bs-dependencies": ["bs-fetch", "reason-future"],
-  "warnings": {
-    "error": "+101"
-  },
-  "refmt": 3
+	"name": "reason-cognito",
```
Comment (pckilgore) [inline bsconfig.json]: Formatter warz.

I'll setup some editorconfig on my branch...

File: src/Cognito.re [hunk 4]
```diff
@@ -83,4 +165,50 @@ let signUp =
         });
       };
     });
+};
+
+open SignInHelper;
+external makeSignInResponse: 't => signInResponseDecoder = "%identity";
+external toJsDict: 'a => Js.Dict.t(Js.Json.t) = "%identity";
```
Comment (pckilgore) [inline src/Cognito.re]: I'd prefer boilerplate dict building over bucklescript magic for something this simple, like: 

https://github.com/pckilgore/reason-cognito/blob/f1b437a8eaafd19abafae9443069537bbce03ec0/src/Cognito.re#L142-L146

File: src/Cognito.re [hunk 5]
```diff
@@ -83,4 +165,50 @@ let signUp =
...
+  /* create and object of this shape using our `authenticationParameters` type
+     ```json
+     {
+       AuthParameters : {
+         "USERNAME": "idkjs",
+         "PASSWORD": "somepassword",
+       }
+     }
+     ```
+     At this point we know we have a valid js object which can test with `Js.Json.test(signInParams, Object)`
```
Comment (pckilgore) [inline src/Cognito.re]: I don't think you logic here is wrong, btw, I just prefer to lean on the type system as much as possible and remove all doubt. Here I think we're talking about 4-5 lines of boilerplate and we're golden.

File: src/Cognito.re [hunk 6]
```diff
@@ -83,4 +165,50 @@ let signUp =
...
+     Since we know its valid we will change its type to Js.Dict.t(Js.Json.t) which is expected by out params dict. We then use `Js.Dict.set(params, "AuthParameters", Js.Json.object_(signInParams));`to push it to the `AuthParameters` key expected by the cognito client.
+      */
+  let params = Js.Dict.empty();
+  Js.Dict.set(params, "AuthParameters", Js.Json.object_(signInParams));
+  Client.request(`InitiateAuth, params)
```
Comment (pckilgore) [inline src/Cognito.re]: I like the variant approach here for the operation!

File: src/Cognito.re [hunk 7]
```diff
@@ -83,4 +165,50 @@ let signUp =
...
+  ->Future.mapOk(resp => {
+      let isErrorResponse = makeResponse(resp);
+      let err = isErrorResponse->__typeGet;
+      let msg = isErrorResponse->messageGet;
+      Js.log2("Raw Response", resp);
+
+      switch (err) {
```
Comment (pckilgore) [inline src/Cognito.re]: See how I'm doing errors now in my most recent dev branch.

I think there's some logic there than can be shared, because all those damn cognito errors makes for a lot of boilerplate.  I think many are shared, though, and since polymorphic variants are composable the mapping operation for any common errors should compose too!

Comment (pckilgore) [reply src/Cognito.re]: FYI, JSCoverter doesn't work there, because I want the variant to carry a payload.

File: src/SignInHelper.re [hunk 8]
```diff
@@ -0,0 +1,35 @@
+
```
Comment (pckilgore) [inline src/SignInHelper.re]: I'm trying to colocate this stuff to the function that consumes it.  If you find yourself needing it more than once, put it in the Types.re file and `open` that.

Eventually I'll have to think harder about whether we modularize the functions here or hide them behind an interface.  But whatever way we go, having them organized as above will make it easier to spot patterns, overlap, and cruft.

File: src/Types.re [hunk 9]
```diff
@@ -0,0 +1,23 @@
+// all operations found in `/amazon-cognito-identity-js/src/CognitoUser.js` except `SignUp` which is in `CognitoUserPool.js`;
+[@bs.deriving jsConverter]
```
Comment (pckilgore) [inline src/Types.re]: :+1:

Comment (pckilgore) [reply src/Types.re]: That comment is really helpful can you add in an actual link to github to make it easier to cross reference?

File: src/Types.re [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  | [@bs.as "ConfirmSignUp"] `ConfirmSignUp
+  | [@bs.as "ChangePassword"] `ChangePassword
+  | [@bs.as "RespondToAuthChallenge"] `RespondToAuthChallenge
+  | [@bs.as "ForgotPassword"] `ForgotPassword
+  | [@bs.as "ConfirmForgotPassword"] `ConfirmForgotPassword
+  | [@bs.as "InitiateAuth"] `InitiateAuth
+];
+/* https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html#API_InitiateAuth_RequestSyntax */
+[@bs.deriving jsConverter]
+type authenticationFlowType = [
```
Comment (pckilgore) [inline src/Types.re]: Nice find, I hadn't see the full list yet.

File: src/Types.re (hunk 9 above)
Comment (pckilgore) [reply src/Types.re]: As in, fully qualified, `https://` ...

File: .gitignore (hunk 2 above)
Comment (idkjs) [reply .gitignore]: Don't know how to do that, actually. Will have to read up on it. That is, git strategies for local work. I will keep the .git file clean by just not adding anything to it!

File: src/Cognito.re (hunk 4 above)
Comment (idkjs) [reply src/Cognito.re]: So this is exactly magic. The result of `deriving abstract` is a valid, testable js object. I did because I couldn't figure out how to get the AuthParameter object into the `params` dict. As I say that, I think I did it somewhere else in this same file so will refactor.

Comment (idkjs) [reply src/Cognito.re]: Done.

File: src/Cognito.re (hunk 6 above)
Comment (idkjs) [reply src/Cognito.re]: Nice. Can implement with auth flow type or convert to the to string method you used in your latest commit. I think `toString` is native compliant so we should go with that. Which do you prefer? 

ToString I am referring to:
https://github.com/pckilgore/reason-cognito/blob/f1b437a8eaafd19abafae9443069537bbce03ec0/src/Types.re#L20-L33

File: src/Cognito.re (hunk 7 above)
Comment (idkjs) [reply src/Cognito.re]: Agree on the repeated error logic. Was thinking through. I am going to get the call we have going first then look to refactor the redundancy if you dont get to it first.

File: src/SignInHelper.re (hunk 8 above)
Comment (idkjs) [reply src/SignInHelper.re]: That file is just getting so long! Hard to scroll between things when working through it. I have actually just been putting the one call, client, everything in one file until it works, the cp into Cognito.re...Is that bad?

File: src/Cognito.re (hunk 6 above)
Comment (pckilgore) [reply src/Cognito.re]: I'm not sure what you're talking about but I think I'm about to learn something good so go for it.

File: src/Cognito.re (hunk 7 above)
Comment (pckilgore) [reply src/Cognito.re]: 👍 IMO duplication is always better than a abstraction you're unsure of.

File: src/Cognito.re (hunk 6 above)
Comment (idkjs) [reply src/Cognito.re]: `bucklescript` isnt supported in native just yet. So we do `bs.whatever` it wont work on native, like this:
```reason
[@bs.deriving jsConverter]
type operation = [
  | [@bs.as "SignUp"] `SignUp
  | [@bs.as "SignIn"] `SignIn
  | [@bs.as "ConfirmSignUp"] `ConfirmSignUp
  | [@bs.as "ChangePassword"] `ChangePassword
  | [@bs.as "RespondToAuthChallenge"] `RespondToAuthChallenge
  | [@bs.as "ForgotPassword"] `ForgotPassword
  | [@bs.as "ConfirmForgotPassword"] `ConfirmForgotPassword
  | [@bs.as "InitiateAuth"] `InitiateAuth
];
```

Where as taking the extra set write a `operataionToString` function, will compile in native.

File: src/SignInHelper.re (hunk 8 above)
Comment (pckilgore) [reply src/SignInHelper.re]: Frankly I tend to build really long files early in a project until I understand what abstractions are appropriate.  I'm not there yet, so I don't really care how long Cognito.re gets!  If you use vscode, you can open it twice to mitigate some of the scrolling pain:
image

File: .gitignore (hunk 2 above)
Comment (idkjs) [reply .gitignore]: I have only rebased once. Not sure how to rebase this pull onto you new api. Any suggestions?

File: src/Cognito.re (hunk 6 above)
Comment (pckilgore) [reply src/Cognito.re]: Ahh I see.  Yeah go for it.  The `toString` way is a bit more boilerplate than the bs helper, but it's super easy to refactor either way so go ahead and do what feels right.

File: .gitignore (hunk 2 above)
Comment (pckilgore) [reply .gitignore]: I guess I say rebase but I guess I don't mean you literally have to rebase.  You can just merge in my changes into this branch.  The "command line instructions" below should explain that.

File: src/SignInHelper.re (hunk 8 above)
Comment (idkjs) [reply src/SignInHelper.re]: I noticed you have added `poolid` and `region` to the paramaters here:
https://github.com/pckilgore/reason-cognito/blob/f1b437a8eaafd19abafae9443069537bbce03ec0/src/Scratch.re

I need a `poolId` and `region` to keep going. Have one? We werent using it before and if I take them out now, it throws:

```
$ node src/Scratch.bs.js
[
  281060686,
  'User pool client 3vjshpa4lgf92nfisjrg9os21a does not exist.'
]
```
Have you tried the new api against the endpoint yet?

Comment (pckilgore) [reply src/SignInHelper.re]: Look at my most recent scratch file?

Comment (pckilgore) [reply src/SignInHelper.re]: I'm at work, but I'll double check when I get home.

Comment (idkjs) [reply src/SignInHelper.re]: nothing new in the most recent scratch in dev branch in the repo. Same dummy params. Will check later when you get a chance to look at it.
https://github.com/pckilgore/reason-cognito/blob/f1b437a8eaafd19abafae9443069537bbce03ec0/src/Scratch.re#L1-L9

**Comment by pckilgore** [comment]
pid: us-east-2_pTFKTGtS3
region: UsEast1

    -  Patrick

**Comment by idkjs** [comment]
Tried it. No go. 
```sh
Raw Response [
  [ 400, tag: 3 ],
  {
    __type: 'ResourceNotFoundException',
    message: 'User pool client 3vjshpa4lgf92nfisjrg9os21a does not exist.'
  }
```

Tried setting up my own but i guess I dont know how you configured it. Maybe share your cognito config later on so we can be working against same config.

**Comment by pckilgore** [comment]
I'm a dumbass, `us-east-2_pTFKTGtS3` is in UsEast2 (duh).

**Comment by idkjs** [comment]
I was like, that's cool, didnt know you could do that, different regions, guess you cant!

File: package.json [hunk 11]
```diff
@@ -5,15 +5,15 @@
     "clean": "bsb -clean-world",
     "build": "bsb -make-world",
     "start": "bsb -make-world -w",
-    "test": "node src/Scratch.bs.js"
+    "test": "node src/scratch.bs.js"
```
Comment (pckilgore) [inline package.json]:
```suggestion
    "test": "node src/Scratch.bs.js"
```

File: package.json [hunk 12]
```diff
@@ -5,15 +5,15 @@
...
   },
   "keywords": [
     "BuckleScript"
   ],
   "author": "",
   "license": "MIT",
   "devDependencies": {
-    "bs-platform": "^5.2.0",
+    "bs-platform": "^5.0.6",
```
Comment (pckilgore) [inline package.json]:
```suggestion
    "bs-platform": "^5.2.0",
```

Comment (pckilgore) [reply package.json]: new bucklescript gets some really nifty compiler optimizations

Comment (pckilgore) [reply package.json]: This will also reduce the diff footprint, because this string gets printed in the generated files.

**Comment by idkjs** [comment]
Details on latest commit are:
`signIn`, `signUp`,`confirmSignUp`,`changePassword` are working. 

I added exceptions as they came up such as here: https://github.com/idkjs/reason-cognito/blob/f394ab580727f4037abab0d31160bf3439e957bf/src/Cognito.re#L135

And here: https://github.com/idkjs/reason-cognito/blob/f394ab580727f4037abab0d31160bf3439e957bf/src/Cognito.re#L135

And here
https://github.com/idkjs/reason-cognito/blob/f394ab580727f4037abab0d31160bf3439e957bf/src/Cognito.re#L212

And here https://github.com/idkjs/reason-cognito/blob/f394ab580727f4037abab0d31160bf3439e957bf/src/Cognito.re#L294

I made a variant of the `operation` type here https://github.com/idkjs/reason-cognito/blob/f394ab580727f4037abab0d31160bf3439e957bf/src/Types.re#L7-L26

and implemented them where applicable.

File: src/Cognito.re [hunk 13]
```diff
@@ -65,10 +65,11 @@ module Client = {
   // returns an error code (4XX, 5XX), that should be handled elsewhere.
   //
   // TODO: More generic implemntation (no Js.* types)
-  let request = (config, operation, params: Js.Dict.t(Js.Json.t)) => {
+  let request = (config, operation: operation, params: Js.Dict.t(Js.Json.t)) => {
```
Comment (pckilgore) [inline src/Cognito.re]: I'd delete this and add this type to the function signature in the module signature above:

```reason
  let request:
    (t, string, Js.Dict.t(Js.Json.t)) =>
    (t, operation, Js.Dict.t(Js.Json.t)) =>
    Future.t(Belt.Result.t(response, [< error]));
    Future.t(Belt.Result.t(response, [< error]));
```

File: bsconfig.json [hunk 14]
```diff
@@ -1,18 +1,23 @@
...
```
Comment (pckilgore) [inline bsconfig.json]: Can you revert all the changes to this file (`git checkout origin/master bsconfig.json` should do it) and push back up?  It looks like the only changes to this file are formatting and I will eventually enforce that in some way for readability/reviewability.

Comment (pckilgore) [reply bsconfig.json]: Actually I have no clue how to do this given somehow github is cool with us having the same branch name.... copy + paste probably easier lol.

File: src/Cognito.re (hunk 13 above)
Comment (pckilgore) [reply src/Cognito.re]: you might have to add `type operation` there too.

File: src/Cognito.re [hunk 15]
```diff
@@ -145,12 +147,15 @@ let signUp =
   Js.Dict.set(payload, "UserAttributes", Js.Json.objectArray(jsonAttribs));
   Js.Dict.set(payload, "ValidationData", Js.Json.objectArray(jsonVData));
 
-  Client.request(config, "SignUp", payload)
-  ->Future.flatMapOk(res =>
+  Client.request(config, SignUp, payload)
+  ->Future.flatMapOk(res => {
+      Js.log2("Raw Response", res);
```
Comment (pckilgore) [inline src/Cognito.re]:
```suggestion
```

File: src/Cognito.re [hunk 16]
```diff
@@ -145,12 +147,15 @@ let signUp =
...
+
       switch (res.status) {
       | Success(_) =>
         // We're _really_ hoping amazon holds to their API contract here.
         // If not, we're going to see null type errors.
+        Js.log2("Raw Response", res);
```
Comment (pckilgore) [inline src/Cognito.re]:
```suggestion
```

File: src/Cognito.re [hunk 17]
```diff
@@ -131,6 +132,7 @@ type signUpErrors = [
   | `CognitoUnexpectedLambda(string)
   | `CognitoUserLambdaValidation(string)
   | `CognitoUsernameExists(string)
+  | `CognitoConfirmationCodeValidation(string)
```
Comment (pckilgore) [inline src/Cognito.re]: Can you link me to where this error is documented?  I didn't see it when I was constructing this...

File: src/Cognito.re [hunk 18]
```diff
@@ -201,6 +206,158 @@ let signUp =
           | _ => `CognitoUnknownError(msg)
           };
         Future.make(resolve => resolve(Belt.Result.Error(err)));
-      }
-    );
+      };
+    });
+};
+type confirmSignUpErrors = [ | `CognitoConfirmationCodeValidation(string)];
```
Comment (pckilgore) [inline src/Cognito.re]: There's a lot more of these we need to handle, see https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmSignUp.html.

File: src/Cognito.re [hunk 19]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+  Js.Dict.set(params, "Username", Js.Json.string(username));
+  Js.Dict.set(params, "ConfirmationCode", Js.Json.string(confirmationCode));
+
+  Client.request(config, ConfirmSignUp, params)
+  ->Future.flatMapOk(res => {
+      /*
+             This is the raw response. Not sure what to do with this if anything. What is `tag:1`
+             Raw Response [ [ 200, tag: 1 ], {} ]
+       Raw Response [ [ 200, tag: 1 ], {} ] */
+      Js.log2("Raw Response", res);
```
Comment (pckilgore) [inline src/Cognito.re]: Delete this and the comment above.  

The response is a Record type, see ~ line 50 above for the definition.  That record field is a variant that carries a payload. The tag corresponds to a variant name, the int to the payload.  So here, this is the compiled version of:
```reason
type status = Success(int)
              ^ Tag1?  ^ Response code 200
```

File: src/Cognito.re [hunk 20]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+      | ClientError(_)
+      | ServerError(_) =>
+        // We're _really_ hoping amazon holds to their API contract here too.
+        // Although the unknownerror variant helps lots.
+        let isErrorResponse = makeResponse(res.json);
+        let err = isErrorResponse->__typeGet;
+        let msg = isErrorResponse->messageGet;
+        let err =
+          switch (err) {
+          | "CodeMismatchException" => `CognitoConfirmationCodeValidation(msg)
```
Comment (pckilgore) [inline src/Cognito.re]:
```suggestion
          | "CodeMismatchException" => `CognitoCodeMismatch(msg)
```
Let's keep the terminology as close to AWS' as possible while stilll being valid reason.  I did drop the extraneous word "exception" here since this is a Result.Error() so we know we should be handling it/ignoring it appropriately.

Comment (pckilgore) [reply src/Cognito.re]: See above, there's a lot more errors documented by AWS at that link, we should handle them all here.  There should be some common cases from SignUp, so feel free to just copy that code and we can worry about an abstraction later.

File: src/Cognito.re [hunk 21]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+          | _ => `CognitoUnknownError(msg)
+          };
+        Future.make(resolve => resolve(Belt.Result.Error(err)));
+      };
+    });
+};
+
+type signInExceptions = [ | `NotAuthorizedException(string)];
```
Comment (pckilgore) [inline src/Cognito.re]: Same deal as above, I imagine there are lots more errors documented and we should handle them all.

File: src/Cognito.re [hunk 22]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+  let params = Js.Dict.empty();
+  Js.Dict.set(params, "AuthParameters", Js.Json.object_(authParams));
+  Js.Dict.set(params, "AuthFlow", Js.Json.string("USER_PASSWORD_AUTH"));
+  Client.request(config, InitiateAuth, params)
+  ->Future.flatMapOk(res => {
+      /*
+             This is the raw response. Not sure what to do with this if anything. What is `tag:1`
+             Raw Response [ [ 200, tag: 1 ], {} ]
+       Raw Response [ [ 200, tag: 1 ], {} ] */
+      Js.log2("Raw Response", res);
```
Comment (pckilgore) [inline src/Cognito.re]: See above, delete.

File: src/Cognito.re [hunk 23]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+      | ServerError(_) =>
+        // We're _really_ hoping amazon holds to their API contract here too.
+        // Although the unknownerror variant helps lots.
+        let isErrorResponse = makeResponse(res.json);
+        let err = isErrorResponse->__typeGet;
+        let msg = isErrorResponse->messageGet;
+        let err =
+          switch (err) {
+          | "InvalidParameterException" => `CognitoInvalidParameter(msg)
+          | "NotAuthorizedException" => `CognitoNotAuthorized(msg)
```
Comment (pckilgore) [inline src/Cognito.re]: See above

File: src/Cognito.re [hunk 24]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+          | _ => `CognitoUnknownError(msg)
+          };
+        Future.make(resolve => resolve(Belt.Result.Error(err)));
+      };
+    });
+};
+let changePassword =
```
Comment (pckilgore) [inline src/Cognito.re]: Let's delete this entire function and work on the stuff above until it's robust, especially since it looks like it is just copy + paste.

Once ConfirmSignup and SignIn feel like they are in a good mergable state, we can move on.  

Besides, I'm allready seeing a lot of patterns (why I like to colocate early), so there might be an abstraction available that will ***significantly*** reduce the amount of code over a boilerplate approach to building these things.

File: src/Types.re [hunk 25]
```diff
@@ -1,7 +1,29 @@
+/* https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_InitiateAuth.html#API_InitiateAuth_RequestSyntax */
 type authenticationFlowType =
   | USER_PASSWORD_AUTH
   | USER_SRP_AUTH
   | CUSTOM_AUTH;
+/* all operations found in https://github.com/aws-amplify/amplify-js/blob/master/packages/amazon-cognito-identity-js/src/CognitoUserPool.js except `SignUp` which is in `CognitoUserPool.js`*/
+type operation =
```
Comment (pckilgore) [inline src/Types.re]: Those docs I liked above have more operations in them we might want to support down the line.  No need to add them now.  Maybe we should start collecting a references library in the README.

File: src/save.txt [hunk 26]
```diff
@@ -1,75 +0,0 @@
-record api:
```
Comment (pckilgore) [inline src/save.txt]: Please restore entire file.

File: yarn.lock [hunk 27]
```diff
@@ -85,10 +85,10 @@ bs-fetch@^0.5.0:
   resolved "https://registry.npmjs.org/bs-fetch/-/bs-fetch-0.5.0.tgz#6913b1d1ddfa0b0a4b832357854e9763d61d4b28"
   integrity sha512-cGjwRpyNcIaX+p2ssy/38zs7BM/miKNgmOR3NEhxKFete5mR05JcvjuV4raG89oGCG281SU1b56TTAKmf9VCug==
 
-bs-platform@^5.2.0:
-  version "5.2.0"
-  resolved "https://registry.yarnpkg.com/bs-platform/-/bs-platform-5.2.0.tgz#0ec317273daed573491c105f68ea48826a285b7a"
-  integrity sha512-miyePsOF9VbuhT5QD5E/hb+l454Fo4MAcg5xV1GJhbWxmejuF/X7mCYUsNrK1UUAaYt8hnoyFdeLG22sxVta9A==
+bs-platform@^5.0.6:
```
Comment (pckilgore) [inline yarn.lock]: Please revert these changes, I want to stay on 5.2.0 (unless you're seeing bugs!).

File: src/Cognito.re [hunk 28]
```diff
@@ -201,6 +206,158 @@ let signUp =
...
+          | "CodeMismatchException" => `CognitoConfirmationCodeValidation(msg)
+          | _ => `CognitoUnknownError(msg)
+          };
+        Future.make(resolve => resolve(Belt.Result.Error(err)));
+      };
+    });
+};
+
+type signInExceptions = [ | `NotAuthorizedException(string)];
+let signIn = (config, ~username: string, ~password: string, ()) => {
```
Comment (pckilgore) [inline src/Cognito.re]: I think the function names should match the corresponding Congito operation, so here, lets rename to `initiateAuth`

signIn, I think , is a abstraction on top of InitiateAuth and other primatives like token exchange.

File: yarn.lock (hunk 27 above)
Comment (idkjs) [reply yarn.lock]: Done.

File: bsconfig.json (hunk 14 above)
Comment (idkjs) [reply bsconfig.json]: You know what happened here? I was working in a folder off git. Then when i got everything going, I just cp'ed into an older version of the fork, which I see now was stupid. Never even thought about it.

File: src/Cognito.re (hunk 17 above)
Comment (idkjs) [reply src/Cognito.re]: That refers to https://docs.aws.amazon.com/cognito-user-identity-pools/latest/APIReference/API_ConfirmSignUp.html. Second error in `Errors` section, "CodeMismatchException". Came up when I was seeing what happened if I put the wrong confirmation code for the code param.

File: src/Cognito.re (hunk 18 above)
Comment (idkjs) [reply src/Cognito.re]: Guess you found it. Just made a note above.

File: src/Cognito.re (hunk 19 above)
Comment (idkjs) [reply src/Cognito.re]: Just to get the habits right, how do you pro devs handle comments in your code when you are working with someone. Apparently, the way I talk to myself in comments doesn't work if anyone is looking at it. 

Thanks for the answer there.

File: src/Cognito.re (hunk 17 above)
Comment (pckilgore) [reply src/Cognito.re]: Ahh, yeah, let's organize that over in a type specific to confirm signup errors, and out of this type.  There will be some duplication but I don't think the type system cares given these are polymorphic variants.  If it does let me know and we can think up some solutions.

File: src/Cognito.re (hunk 18 above)
Comment (pckilgore) [reply src/Cognito.re]: Ahh, I actually didn't make the connection.  This is good down here, remove it from the type above. :+1:

File: src/Cognito.re (hunk 19 above)
Comment (pckilgore) [reply src/Cognito.re]: Ahh, there really are no 100% rules here.  Feel free to keep doing this in the future, I'm happy to answer the questions posed, but I like to keep the codebase clean of this sort of thing before merger.


Personally I tend to have a lot of comments to myself as I work through a new problem, then delete most of them as I figure out the approach.  The only ones I leave are explanatory, especially ones that explain _why_ I made a choice to do something a certain way.  Or to document as assumption I'm making that's not explicit.  I don't think there's a right answer??

Here, I (hope I) answered your question, so I figured let's kill the comment?

File: package.json (hunk 12 above)
Comment (idkjs) [reply package.json]: Done.

File: package.json (hunk 11 above)
Comment (idkjs) [reply package.json]: Done.

File: src/Cognito.re (hunk 13 above)
Comment (idkjs) [reply src/Cognito.re]: Done. Compiler picks up `Types.operation` so did not have to annotate.

File: src/Cognito.re (hunk 24 above)
Comment (idkjs) [reply src/Cognito.re]: Lead the way, sir.

File: src/Types.re (hunk 25 above)
Comment (idkjs) [reply src/Types.re]: I looked at some of those operations. Some of them just take unit and no params so might have to make the params on the client request type optional. I am thinking out loud here. I am not getting how to call some of these operations. For example, if we want to call `getCurrurentContextData` found here:https://github.com/aws-amplify/amplify-js/blob/2fbfde04092c2b0bfddc3d9698975ba7e9c5e572/packages/amazon-cognito-identity-js/src/CognitoUserPool.js#L156

I tried this:
```reason
let getUserContextData = (config, ~username: string, ()) => {

  let params = Js.Dict.empty();
  Js.Dict.set(params, "USERNAME", Js.Json.string(username));
  Client.request(config, InitiateAuth, params)
  ->Future.flatMapOk(res =>
      switch (res.status) {
      | Success(_) =>
        Js.log2("RES", res);
        Future.make(resolve =>
          resolve(
            Belt.Result.Ok(
              {
                res;
              },
            ),
          )
        );

      | Informational(_)
      | Redirect(_)
      | ClientError(_)
      | ServerError(_) =>

        let isErrorResponse = makeResponse(res.json);
        let err = isErrorResponse->__typeGet;
        let msg = isErrorResponse->messageGet;
        let err =
          switch (err) {
          | "InvalidParameterException" => `CognitoInvalidParameter(msg)
          | "NotAuthorizedException" => `CognitoNotAuthorized(msg)
          | _ => `CognitoUnknownError(msg)
          };
        Future.make(resolve => resolve(Belt.Result.Error(err)));
      }
    );
};
```

Then call with:

```reason
// Scratch.re
let getUserContextData =
  auth->Cognito.getUserContextData(~username="sometestuser2");
```

The function runs and does nothing. What does that mean? Also, the function takes no params but obviously it must because it needs a username. How would I read the signature in the linked function in order to call it? What operation does that function expect to receive? Teach, brother, teach.

**Comment by pckilgore** [comment]
I wasn't feeling that great yesterday and now I'm a little behind at work so I'll get to this ASAP, might be until the weekend though.  Thanks so much!

**Comment by idkjs** [comment]
Your back! Nice. I am all in, boss. Will look for the blue dot.
